src/test/java/org/mockito/internal/stubbing/defaultanswers/ReturnsSmartNullsTest.java

```diff
@@ -85,4 +90,36 @@ public void should_print_the_parameters_on_SmartNullPointerException_message() t
                 .hasMessageContaining("lumpa");
         }
     }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void generic_return_type_erasure() throws Throwable {
+        class S<T> {
+            public T f() {
+                return null;
+            }
+        }
+
+        S<Boolean> s = Mockito.mock(S.class, org.mockito.Answers.RETURNS_SMART_NULLS);
+
+        Mockito.when(s.f()).thenReturn(true);
+        assertTrue(s.f());
+    }
+
+    // https://github.com/mockito/mockito/issues/357
+    @Test
+    public void inferred_generic_type() throws Throwable {
+        class F {
+            <T> T g(Iterable<T> xs) {
```

Comment:
I do not think that we can even fix this. The inferred type is not available from within the method. This would require to parse the byte code from the method's call site and even with advanced techniques like rebinding the call site to invokedyanmic, it is not always possible to decide the correct type for the smart nulls.